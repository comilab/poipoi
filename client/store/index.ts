import { defineComponent, provide, inject } from 'nuxt-composition-api'
import { InjectionKey } from '@vue/composition-api'

import sessionStore from '~/store/session'
import settingStore from '~/store/setting'
import toastStore from '~/store/toast'

type Store = {
  session: ReturnType<typeof sessionStore>
  setting: ReturnType<typeof settingStore>
  toast: ReturnType<typeof toastStore>
}

const makeStore = () => {
  return {
    session: sessionStore(),
    setting: settingStore(),
    toast: toastStore()
  }
}

export const key: InjectionKey<Store> = Symbol('Store')

export const store: Store = makeStore()

export const StoreProvider = defineComponent({
  setup () {
    provide(key, store)
  },
  render (h) {
    return h('div', this.$slots.default)
  }
})

export const useStore = () => inject(key)!

export const resetStore = () => {
  Object.assign(store, makeStore())
}

export const actions = {
  async nuxtClientInit () {
    try {
      await Promise.all([
        store.session.load(),
        store.setting.load()
      ])
    } catch (error) {
      //
    }
  }
}
