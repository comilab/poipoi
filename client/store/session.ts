import { reactive, toRefs, computed } from 'nuxt-composition-api'
import axios from 'axios'

import User from '~/models/User'

export default function sessionStore () {
  const state = reactive({
    user: null as User|null
  })

  const isLogin = computed(() => !!state.user)

  async function create (query: any) {
    await axios.get('/sanctum/csrf-cookie', {
      baseURL: process.env.apiUrl
    })
    await axios.post('/sessions', {
      ...query,
      device_name: 'poipoi-web'
    })
  }

  async function load () {
    try {
      const { data } = await axios.get('/sessions')
      setUser(new User(data))
    } catch (error) {
      setUser(null)
    }
  }

  async function destroy () {
    try {
      await axios.delete('/sessions/poipoi-web')
    } catch (error) {
      //
    }
    setUser(null)
  }

  function setUser (user: User|null) {
    state.user = user
  }

  return {
    ...toRefs(state),
    isLogin,
    create,
    load,
    destroy,
    setUser
  }
}
