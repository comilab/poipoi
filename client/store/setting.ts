import { reactive, toRefs } from 'nuxt-composition-api'
import axios from 'axios'

import useApi from '~/composables/use-api'
import Setting from '~/models/Setting'

export default function sessionStore () {
  const api = useApi('/settings', Setting)

  const state = reactive({
    data: null as Setting|null
  })

  async function load () {
    const { data } = await axios.get('/settings')
    setData(new Setting(data))
  }

  async function save (data: any) {
    await api.save(data)
    setData(api.item.value!)
  }

  function setData (data: Setting|null) {
    state.data = data
  }

  return {
    ...toRefs(state),
    load,
    save,
    setData
  }
}
