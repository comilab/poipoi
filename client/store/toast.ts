import { reactive, toRefs } from 'nuxt-composition-api'

export type Toast = {
  type: 'success'|'info'|'warning'|'danger'
  message: string
  timeout?: number
}

export default function toastStore () {
  const state = reactive({
    toasts: [] as Toast[]
  })

  function push ({ type, message, timeout = 5000 }: Toast) {
    const toast = {
      type,
      message,
      timeout
    }
    state.toasts.push(toast)
    if (timeout) {
      setTimeout(() => {
        remove(toast)
      }, timeout)
    }
  }

  function remove (toast: Toast) {
    const index = state.toasts.findIndex(t => t === toast)
    if (index >= 0) {
      state.toasts.splice(index, 1)
    }
  }

  return {
    ...toRefs(state),
    push,
    remove
  }
}
