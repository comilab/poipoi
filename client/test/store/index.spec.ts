import { mocked } from 'ts-jest/utils'
import { render } from '@testing-library/vue'
import { defineComponent } from 'nuxt-composition-api'

import { store, StoreProvider, useStore, actions } from '~/store'

describe('store/index', () => {
  test('store', () => {
    expect(store.session).toBeDefined()
    expect(store.setting).toBeDefined()
    expect(store.toast).toBeDefined()
  })

  test('StoreProvider, useStore', () => {
    let injectedStore

    const inner = defineComponent({
      setup () {
        injectedStore = useStore()
      },
      template: '<div>inner</div>'
    })

    const wrapper = defineComponent({
      components: { StoreProvider, inner },
      template: '<StoreProvider><inner /></StoreProvider>'
    })

    const { container } = render(wrapper)

    expect(container).toMatchSnapshot()
    expect(injectedStore).toBe(store)
  })

  describe('actions', () => {
    describe('nuxtClientInit', () => {
      let nuxtClientInit: typeof actions.nuxtClientInit

      beforeEach(() => {
        nuxtClientInit = actions.nuxtClientInit
        jest.spyOn(store.session, 'load').mockResolvedValue()
        jest.spyOn(store.setting, 'load').mockResolvedValue()
      })

      test('calls store.session.load', async () => {
        await nuxtClientInit()

        expect(store.session.load).toBeCalled()
      })

      test('calls store.setting.load', async () => {
        await nuxtClientInit()

        expect(store.setting.load).toBeCalled()
      })

      test('エラーが起きても何もしない', () => {
        mocked(store.session.load).mockRejectedValue(new Error())

        expect(async () => {
          await nuxtClientInit()
        }).not.toThrow()
      })
    })
  })
})
