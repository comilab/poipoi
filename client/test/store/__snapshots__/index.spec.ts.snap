// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`store/index StoreProvider, useStore 1`] = `
<div>
  <div>
    <div>
      inner
    </div>
  </div>
</div>
`;
